Contributions/Linux_Memory_Management_Essentials.md

```diff
@@ -0,0 +1,167 @@
+Individual points are numbered for ease of reference, but the numbering is not meant to represent any sequence.
+
+## **Memory management in Linux**
+
+### **Kernel-space memory allocations**
+
+#### **Facts**
+1. unlike processes memory, kernel memory pages are not swapped, nor dropped silently by the kernel itself,
+   although an hypervisor will do to a VM what the kernel does to a process (but this is beyond the control of the kernel)
+2. the kernel context (usually EL1 on ARM64) uses one single memory map (page tables) across all the cores
+   executing in kernel mode
+3. on 64 bit systems (e.g. ARM64 and x86_64), usually almost all physical memory is mapped in a
+   (semi)contiguous (there can be holes) range. Memory within this range is both virtually and physically contiguous.
+4. physically contiguous memory is treated as a scarce resource, and typically is not provided to userspace,
+   unless it explicitly asks for it (e.g. for implementing some DMA buffer)
+5. the kernel can access userspace in two ways:
+   1. through the userspace mapping
+      1. it is done in exceptional cases, like copy_to_user_copy_from_user
+      2. it uses the userspace memory map, which can implement HW protections against kernel misuses
+      3. it allows the kernel to see process pages in the same sequence and with the same mappings as the proces does
+   2. through the EL1 mappings + linear map
+      1. not the intended way to access process context
+      2. not very useful, even for an attacker, in a security scenario
+      3. the sequence of pages mapped in the userspace process is not known
+      4. some process pages might even be "missing", because they have been either swapped out or dropped
+      5. bypasses any protection that the process might employ through its own mapping
+6. attempt to keep pages hot in the system cache make it very difficult to predict which pages might surface in a certain place:
+   1. during various operations, it is necessary to reserve one or more pages, that later on will be released.
+   2. the kernel adopts various optimisations for attempting to keep low the fragmentation, thus allowing
+      for availability of higher-order allocations, though slabs/buddy allocator and folios.
+   3. Even the slab allocator contains queues of free chunks, grouped by size
+   4. However, it also attempts to cache locally these free pages, in per-cpu queues, with the intent of minimising
+      the amount of cache flushes that would be driven by releasing a recently feed page, and subsequently allocating
+      another one
+
+
+### **User-space memory allocations**
+
+#### **Facts**
+1. user-space memory must be assumed to not always be backed by physical memory pages
+   meaning that a non running process might have had some of its pages dropped or swapped out to disk
+2. unless some extra effort has been taken, to pin down memory allocations for user-space, it is
```

Comment:
It was kind of stating what you pointed out in your previous comment, that even a running process could have some pages dropped/swapped, unless it has taken some care to prevent it.